pdp-server/src/api/authn_middleware.rs

```diff
@@ -0,0 +1,172 @@
+use crate::state::AppState;
+use axum::{
+    body::Body,
+    extract::{Request, State},
+    http::StatusCode,
+    middleware::Next,
+    response::Response,
+};
+use log::warn;
+
+pub(super) async fn authentication_middleware(
+    State(state): State<AppState>,
+    request: Request<Body>,
+    next: Next,
+) -> Response {
+    // Extract the authorization header
+    let auth_header = match request.headers().get(http::header::AUTHORIZATION) {
+        Some(header) => header,
+        None => {
+            warn!("Missing Authorization header");
+            // TODO avoid this expect panic (maybe using IntoResponse)
+            return Response::builder()
+                .status(StatusCode::UNAUTHORIZED)
+                .body("Missing Authorization header".into())
+                .expect("Failed to create response");
+        }
+    };
+
+    // Extract the token from the authorization header
+    let api_key = match auth_header.to_str() {
+        Ok(header_str) if header_str.to_lowercase().starts_with("bearer ") => {
+            // Remove the "Bearer " prefix
+            header_str[7..].to_string()
+        }
+        Ok(header_str) => {
+            warn!(
+                "Invalid Authorization header format, missing 'Bearer ' prefix: {}",
+                header_str
+            );
+            return Response::builder()
+                .status(StatusCode::FORBIDDEN)
+                .body(
+                    "You are not authorized to access this resource, please check your API key."
+                        .into(),
+                )
+                .expect("Failed to create response");
+        }
+        Err(e) => {
+            warn!("Failed to parse Authorization header to string: {}", e);
+            return Response::builder()
+                .status(StatusCode::FORBIDDEN)
+                .body(
+                    "You are not authorized to access this resource, please check your API key."
+                        .into(),
+                )
+                .expect("Failed to create response");
+        }
+    };
+
+    // Verify the API key
+    if api_key != state.settings.api_key {
+        warn!("Authentication failed: Invalid API key");
+        return Response::builder()
+            .status(StatusCode::FORBIDDEN)
+            .body(
+                "You are not authorized to access this resource, please check your API key.".into(),
+            )
+            .expect("Failed to create response");
+    }
+    next.run(request).await
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test::{create_test_state, setup_test_settings};
```

Comment:
Is this a common thing to do in rust ? To create a shared test utils inside a module named test ?

Reply:
No, usually you add methods to the impl and wrap them under `#[cfg(test)]`. I've slowly migrated some functions to this mechanism, the state is now removed